### DISCUSSION TITLE
New WebAssembly backend

### DISCUSSION THREAD

**Discussion by konsoletyper** [konsoletyper/teavm#963] [Announcements]
Hi, I'm glad to announce that I just published preview of new WebAssembly GC backend.

Let me dig dive in history a little bit. First WebAssembly backend was ready somewhere around 2016/2017, when WebAssembly MVP spec was in development and browsers still had bugs in Wasm support and could not even print disassembly. This old backend works perfectly, but I never had enough motivation to publish some docs and encourage community to switch to WebAssembly. The main reason was the inability to create seamless interoperation layer between Java and JS objects.

This is caused by the fact that WebAssembly does not (or at least did not) operate in terms of objects, but rather has a large flat piece of memory called "heap". TeaVM Wasm BE has its own allocator and its own tracing GC implementation, whereas JS lives with its own GC. And it's the fact: one can't just make two different managed heaps each with own GC to communicate to each other. Or at least, communicate seamlessly. This means, that from both sides objects needed to be explicitly "acquired" and "released", which would lead to some sort or reference counting, even when one want to pass callback to JS API.

Another issue was the size of generated binary and its performance. Performance is better than that of JavaScript, but not significantly better. But the generated binary is huge, usually 5x times larger than JS generated from the same source. On one hand, this could be optimized by shrinking data section, which had just unpacked representation of Java class metadata, by providing some sort of compression. On the other hand, I could do nothing with code section.

Why did these issues occur? Because of GC! No, GC itself is tiny, about 10K, but the problem with WebAssembly is that it does not allow to access CPU stack. Usually, managed runtimes access stack directly when they perform mark phase of GC, but in the absence of access to the stack the only viable alternative is maintaining shadow stack, which in turn has tremendous overhead, both size- and performance-wise.

[WebAssembly GC proposal](https://github.com/WebAssembly/gc) adds objects (or structures in its terms) to WebAssembly, so it becomes possible to make references between JS and WebAssembly objects, which effectively allows to re-create [JSO](https://teavm.org/docs/runtime/jso.html), Java-to-JavaScript interaction layer. This in turn allows to take existing programs and almost without modifications compile them with new backend. Chrome and Firefox added support for this proposal almost a year ago, Safari provides it as technology preview.

I spent several months implementing support for WebAssembly GC. This is not just "small extension", this proposal is a game changer that requires to re-think and re-write entire backend, which I did. Instead of modifying existing WebAssembly BE, I just created a new one, so both co-exist together. My experiments have shown that with Wasm GC performance is a bit better than with Wasm MVP and much better than with JS (although still far from near native speed), while binary size is comparable to JS. Given this, and given that WebAssembly has native int64 support (which allows fast `long` implementation) and float32 (which allows to avoid some rare issues with rounding in case of using `float`), I'm determined to push new Wasm GC BE to release stage and to encourage community to switch to it.

As for now, Wasm GC BE supports almost everything the JS BE does, except for:

* "Advanced" reflection, i.e. traversing annotations, traversing and calling methods and fields of classes, getting class by name
* Support for `@Async` annotation, which means not only support of this annotation in user code, but also those methods in TeaVM classlib that utilize `@Async` internally, like `Thread.sleep`, `Object.wait`, etc.
* IDE support
* Development server
* Passing arrays to and from JS by reference. Note that this isn't just "unimplemented" feature, but the thing that impossible to implement even in theory, unless WebAssembly adds ability to represent Wasm GC arrays as JS `ArrayBuffer`. I'm planning to introduce extension to JSO to pass NIO buffers between JS and Java, which would allow to distinguish same semantics.
* Full ESM support, since as of now browsers don't integrate WebAssembly modules in ES module system, but wrappers can be written manually and eventually TeaVM could generate such wrappers.

To try WebAssembly GC in action, you can open [old good sample](https://teavm.org/gallery/jbox2d/index.html). Unfortunately, I still did not have a chance to run WebAssembly GC with something really big, only some small examples like this.

If you want to try WebAssembly GC as a developer, you better ask me directly or learn code, there's still no documentation at all. As a quick start with gradle, learn [this sample](https://github.com/konsoletyper/teavm/tree/7a2cf555e3d0eb05e25c7eb71eed21aa1db8176a/samples/benchmark). In short, everything related to Wasm GC is located in `teavm.wasmGC` DSL. And don't forget that new backend is only available in latest [preview build](https://teavm.org/docs/intro/preview-builds.html).

To start with Maven, you need to set `tagetType` property of your plugin execution to `WEBASSEMBLY_GC`. Additionally, you need to configure copying of WasmGC runtime:

```xml
          <execution>
            <id>please-specify-some-appropriate-id</id>
            <goals>
              <goal>copy-webassembly-gc-runtime</goal>
            </goals>
            <configuration>
              <targetDirectory>${project.build.directory}/directory/to/copy/runtime/to</targetDirectory>
            </configuration>
          </execution>
```

From HTML side you need something like this:

```html
<!DOCTYPE html>
<html>
<head>
    <title>Main page</title>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8">
    <script type="text/javascript" charset="utf-8" src="path/to/fileName.wasm-runtime.js"></script>
</head>
<body>
    <script>
        async function main() {
            let teavm = await TeaVM.wasmGC.load("path/to/fileName.wasm");
            teavm.exports.main([]);
        }
        main();
    </script>
</body>
</html>
```

During next few months I'm going to prepare backend for release, which *does not mean* that I'm going to implement missing features, but rather fix bugs and write some documentation. I hope for your feedback and good luck with trying TeaVM.

**Comment by Ihromant** [comment]
Magnificent job, Alexey!

I can try my pet project with WASM GC. It has 40k lines of code, interoperates with DOM. It matches your description (I assume that I'm using nothing you mentioned in "except"). Project currently compiles with maven. Probably there should be some kind of FAQ or thread for people who are trying WASM GC and facing unknown issues.

**Reply by konsoletyper** [reply]
> Probably there should be some kind of FAQ or thread for people who are trying WASM GC and facing unknown issues.

I suppose, this can be the thread for the first time. Feel free to ask questions right here in comments

**Reply by Ihromant** [reply]
Stupid question probably, but received following error:
Failed to copy webassembly runtime: /path/to/project/webview-local/target/wasm/classes.wasm-runtime.js
Added execution snippet to teavm plugin executions before compile execution.

**Reply by konsoletyper** [reply]
Yes, it's a known issue, already fixed in master. Just create `/path/to/project/webview-local/target/wasm/` manually.

**Reply by Ihromant** [reply]
Fixed compilation, it was successful.
Configured everything, got following error:
```
Uncaught (in promise) CompileError: WebAssembly.compileStreaming(): Compiling function #1801 failed: call[2] expected type externref, found local.get of type (ref null 2541) @+393077
```
in browser console.

**Reply by konsoletyper** [reply]
Looks like TeaVM has generated something wrong. Can you share your projects so that I tried to reproduce this on my side?

**Reply by Ihromant** [reply]
This project is pretty complex. I can show you on video how to launch it, then you'll have a playground to test.

**Reply by konsoletyper** [reply]
Another way is following. First, you need to compile without minification (this will turn generation of `name` segment). Second, you need to generate disassembly, which is a bit complex in case of Maven. The disassembler is available in `teavm-core` artifact as `org.teavm.backend.wasm.disasm.Disassembler` main class, so if you run something like `java -cp proper:class:path org.teavm.backend.wasm.disasm.Disassembler`, you'll able to get disassembly. Then, to have proper classpath, you need `teavm-core.jar`, as well as few dependencies. I guess, `teavm-relocated-libs-hppc` will be enough. So, generate disassembly (there's also option to generate HTML with addresses). Disassembly should help you to identify place which causes issue. You either perform these steps, perhaps referring to source code and get proper disassembly, or rather share your project.

**Reply by Ihromant** [reply]
I don't want to make my project public now. And project in current state is kind of "works on my machine". Generally I could try to migrate to Gradle (I'm using simple dependencies fetching and teavm right now only) if it simplify disassembly generation.

**Reply by xpenatan** [reply]
Hi, trying to get basic libgdx demo working with webassembly. I also get externref error if I pass a JSObject interface like EventListener to a method or static method inside a normal class.  If I remove "extend JSObject" from the interface it runs fine.

```
public static void test (EventListener<?> listener) {
}

// Call
test(new EventListener<Event>() {
    @Override
    public void handleEvent(Event evt) {
    }
});
```

**Reply by konsoletyper** [reply]
@xpenatan thanks, I'll take a look

**Reply by konsoletyper** [reply]
> I don't want to make my project public now.

You can send it to me directly. Also, I decided to publish disassembler as a separate tool, I'll let you know when it's ready

**Reply by Ihromant** [reply]
Sent you an invitation to collaborate. Project can be run in browser through https://github.com/Ihromant/ihromant.github.io develop branch.

**Reply by Ihromant** [reply]
Also it depends on https://github.com/ihromant/teavm-io

**Reply by konsoletyper** [reply]
@Ihromant and in the meantime: https://teavm.org/tools/wasm/disasm/

Also, you can add to your `pom.xml` to TeaVM configuration following:

```xml
<debugInformationGenerated>true</debugInformationGenerated>
<wasmDebugInfoLocation>EMBEDDED</wasmDebugInfoLocation>
<minifying>false</minifying>
```

and the produced disassembly will show all symbols and even references to original source locations

**Reply by konsoletyper** [reply]
> If I remove "extend JSObject" from the interface it runs fine.

Yes, but this changes semantics. I was able to reproduce this issue in tests, so will fix soon.

**Reply by konsoletyper** [reply]
> If I remove "extend JSObject" from the interface it runs fine.

Fixed in master.

**Reply by konsoletyper** [reply]
@Ihromant I tried to build your project and identified two issues (one of which was also reported by @xpenatan). After fixing these issues I managed to get wasm module that is accepted by browsers. Since I have no idea how to run your project, that was enough for me. Can you please build latest TeaVM locally and test your project with it?

**Reply by Ihromant** [reply]
Sure, I will try to do it

**Reply by konsoletyper** [reply]
@Ihromant oh, I just figured out how to run it. Managed to register account, login and start game, but then can't move any rider. Anyway, it looks like TeaVM GC now works for your scenario. I'll wait for feedback from @xpenatan and if everything ok from his side, I'll publish new preview

**Reply by Ihromant** [reply]
Wow, nice. If you were able to start game, then IO layer at least works together with DOM operations. When you publish preview, I'll check again.

**Reply by konsoletyper** [reply]
@Ihromant please, check ASAP, don't wait for publication. Is there anything difficult to test with master?

**Reply by Ihromant** [reply]
Strange, game starts, but I don't see heroes and map (see screenshot).
![Screenshot_2024-10-23_23-22-52](https://github.com/user-attachments/assets/5b933d99-a6b3-4660-9060-84f01440c227)
Stacktrace:
```
83classes.wasm-runtime.js:1 Uncaught RuntimeError: dereferencing a null pointer
    at classes.wasm:0x35eae
    at classes.wasm:0x3f727
    at classes.wasm:0xe76dc
    at classes.wasm:0xe697c
    at classes.wasm:0xe0ab7
    at classes.wasm:0xe0ac8
    at classes.wasm:0x5f863
    at HTMLComponent$addEventListener$lambda$_9_0.eval [as handleEvent] (eval at defineMethod (classes.wasm-runtime.js:1:5143), <anonymous>:5:16)
    at HTMLDivElement.r (classes.wasm-runtime.js:1:6087)
```

**Reply by konsoletyper** [reply]
> Stacktrace:

Please, I'm a good engineer, but not a god, I'm only human like you. Set `minifying` to false, as I described above. This will at least print readable function names. Also, if you generate source maps, you'll get files and line numbers.

**Reply by Ihromant** [reply]
Sorry, I assumed that you could reproduce this NPE locally. Set minifying to false. Still, source maps are not included despite I put them near wasm file.
![Screenshot_2024-10-24_09-44-29](https://github.com/user-attachments/assets/4dbde0d9-ba1d-4533-890d-f754632ae39f)
```
Uncaught RuntimeError: dereferencing a null pointer
    at ua.ihromant.map.proxy.TileProxy::terrain (TileProxy.java:12:1)
    at ua.ihromant.map.proxy.MoveMap::dijkstra (MoveMap.java:72:1)
    at ua.ihromant.sod.client.controller.MapDisplayLogic::findCursor (MapDisplayLogic.java:64:1)
    at ua.ihromant.sod.client.controller.MapController::handleMove (MapController.java:297:1)
    at ua.ihromant.sod.client.controller.MapController$<init>$lambda$_0_6::accept (MapController.java:105:1)
    at ua.ihromant.sod.client.controller.MapController$<init>$lambda$_0_6::accept@caller (classes.wasm:0xe0ddc)
    at ua.ihromant.sod.view.composite.HTMLComponent$addEventListener$lambda$_9_1::handleEvent$exported$0 (HTMLComponent.java:113:1)
    at HTMLComponent$addEventListener$lambda$_9_1.eval [as handleEvent] (eval at defineMethod (classes.wasm-runtime.js:1:5143), <anonymous>:5:16)
    at HTMLDivElement.r (classes.wasm-runtime.js:1:6087)
```
Just checked, in JS everything works fine.

**Reply by Ihromant** [reply]
There is another error that prevents game from initial rendering:
![Screenshot_2024-10-24_10-14-07](https://github.com/user-attachments/assets/e7477694-6e2e-46b4-84bc-609f634ff5e4)
```
Uncaught RuntimeError: dereferencing a null pointer
    at ua.ihromant.map.proxy.TileProxy::terrain (TileProxy.java:12:1)
    at ua.ihromant.sod.client.controller.MapController::repaintPixel (MapController.java:413:1)
    at ua.ihromant.sod.client.controller.MapController::repaintMinimap (MapController.java:396:1)
    at ua.ihromant.sod.client.state.AppState$fillGame$lambda$_12_0::accept (Coordinate.java:9:1)
    at ua.ihromant.sod.client.state.AppState$fillGame$lambda$_12_0::accept@caller (classes.wasm:0xb3d17)
    at ua.ihromant.sod.shared.asset.AssetFetcher$1::assetLoaded (AssetFetcher.java:44:1)
    at ua.ihromant.sod.shared.asset.AssetFetcher::forAssets (AssetState.java:21:1)
    at ua.ihromant.sod.client.state.AppState::fillGame (AppState.java:262:1)
    at ua.ihromant.sod.client.state.AppState$onMessage$lambda$_2_1::accept (AppState.java:110:1)
    at ua.ihromant.sod.client.state.AppState$onMessage$lambda$_2_1::accept@caller (classes.wasm:0x6b816)
```
Still, as the root is the same, I assume that they`re related.

**Reply by konsoletyper** [reply]
@Ihromant fixed in master. Now I'm able to play, but sometimes I get errors like this:

```
java.lang.IllegalArgumentException: Bad action SelectObject[object=ObjectId[type=hero, id=26]]
  at Controller.applyAction(Controller.java:40)
  at MapController.onSelectHero(MapController.java:316)
  at MapController$<init>$lambda$_0_1.accept(MapController.java:96)
  at HeroSelectionView.lambda$new$1(HeroSelectionView.java:32)
  at HeroSelectionView$lambda$new$2$lambda$_10_1.accept(HeroSelectionView.java:31)
  at Optional.ifPresent(TOptional.java:75)
```

for example, when I try select heroes, move them to certain locations on map, etc.

**Reply by Ihromant** [reply]
@konsoletyper It's WIP, i'm constantly developing it and adding new functionality. Readiness is approx 60-70% (at least I hope, I already did 2 wrong estimations :) ). This is (probably) expected, so thanks for your assistance. Will check.

**Reply by Ihromant** [reply]
Wow, now I can confirm that it works as expected. This is my confirmation. Now waiting for @xpenatan

**Reply by xpenatan** [reply]
The extend JSObject is fixed here but Im having another issue related to a hack I did that worked for javascript and does not with  webassembly hehe 😆.  I guess @konsoletyper can give me some tips.

Libgdx desktop GL classese use Buffer, IntBuffer, FloatBuffer and some more to call GL commands.  The problem is that webgl uses Float32Array, Int32Array, Int8Array for GL calls. The early days with gdx-backend I was just copying every byte and creating a new typed array every frame which is a big performance hit.

Few months ago I did some hacks to improve performance in Javascript side to obtain the typedarray data from teavm "T" class like TShortBufferOverArray, TIntBufferOverArray, TFloatBufferOverByteBuffer and perfornace was improved.  Is there a way to obtain typed arrays from java buffers ?

**Reply by konsoletyper** [reply]
@xpenatan on JS side you don't need any hacks with `T` classes. You can just use existing Java API and call `array` method on buffers. Then you can define native method that takes Java array by reference and returns corresponding JS TypedArray. Perhaps, I should have just included these converters right into TypedArray wrappers in JSO out-of-the box.

As for Wasm GC side, please, read my post. I mentioned this: Wasm GC arrays aren't available as JS `ArrayBuffer`. So there's no way to use `@JSByRef` in Wasm GC. I'm going to slightly different thing: implement Java NIO buffers over Wasm heap (which is available as `ArrayBuffer`) and this will make it possible to pass direct NIO buffers to/from JS side without copying. But this will be later. Anyway, I would appreciate if you made some experiments, even at cost of performance loss on copying data to WebGL. If you make it right now, it will make life easier for you and me when I finally implement buffers.

**Reply by xpenatan** [reply]
There are some cases that libgdx creates a TIntBufferOverByteBuffer object instance (TIntBufferOverByteBufferBigEndian or TIntBufferOverByteBufferLittleEndian, dont remember now) somewhere in their code and was not possible to obtain the javascript array because the getArray throws an exception.  The hack I did was to read byteByffer internally and obtain the int8Array.  Having a helper to obtain int8Array will help a lot.

**Comment by xpenatan** [comment]
* Libgdx tests:

**Reply by xpenatan** [reply]
I skiped the externref error part to test more stuff. 

I get builld error when creating a ModelBatch instance

I have no idea where inside ModelBatch is causing it.

![image](https://github.com/user-attachments/assets/5a1f5a1f-7de4-4d35-8806-2f47dc33dcae)

**Reply by konsoletyper** [reply]
@xpenatan have not idea what you doing. Can you please provide more details? And from this screenshot I understand *nothing*. Also, it's strange that you have mentions of `org.teavm.interop.Address`. Are you sure you configured *wasm GC*, not *wasm*?

**Reply by konsoletyper** [reply]
@xpenatan, also, it's a good idea to start using `org.teavm.tooling.TeaVMProblemRenderer`

**Reply by xpenatan** [reply]
I'm trying to run a very simple 3d example. I found out that ModelBatch class is failing at the compilation step, I'm using TeaVMTargetType.WEBASSEMBLY_GC.

Nice, didnt know about TeaVMProblemRenderer. it gives more details!

```
Method is not annotated with org.teavm.interop.Import
    at org.teavm.interop.Address.getByte
    at java.nio.ByteOrder.nativeOrder(TByteOrder.java:45)
    at com.badlogic.gdx.utils.BufferUtils.newIntBuffer(BufferUtilsEmu.java:373)
    at com.badlogic.gdx.graphics.g3d.utils.DefaultTextureBinder.getMaxTextureUnits(DefaultTextureBinder.java:73)
    at com.badlogic.gdx.graphics.g3d.utils.DefaultTextureBinder.<init>(DefaultTextureBinder.java:62)
    at com.badlogic.gdx.graphics.g3d.utils.DefaultTextureBinder.<init>(DefaultTextureBinder.java:58)
    at com.badlogic.gdx.graphics.g3d.ModelBatch.<init>(ModelBatch.java:82)

Method is not annotated with org.teavm.interop.Import
    at org.teavm.interop.Address.ofData
    at java.nio.ByteOrder.nativeOrder(TByteOrder.java:45)
    at com.badlogic.gdx.utils.BufferUtils.newIntBuffer(BufferUtilsEmu.java:373)
    at com.badlogic.gdx.graphics.g3d.utils.DefaultTextureBinder.getMaxTextureUnits(DefaultTextureBinder.java:73)
    at com.badlogic.gdx.graphics.g3d.utils.DefaultTextureBinder.<init>(DefaultTextureBinder.java:62)
    at com.badlogic.gdx.graphics.g3d.utils.DefaultTextureBinder.<init>(DefaultTextureBinder.java:58)
    at com.badlogic.gdx.graphics.g3d.ModelBatch.<init>(ModelBatch.java:82)
    at com.badlogic.gdx.graphics.g3d.ModelBatch.<init>(ModelBatch.java:125)
```
So the issue is calling ByteOrder.nativeOrder().

**Reply by konsoletyper** [reply]
@xpenatan 

> ```
> Method is not annotated with org.teavm.interop.Import
>     at org.teavm.interop.Address.getByte
>     at java.nio.ByteOrder.nativeOrder(TByteOrder.java:45)
> ```

Fixed in master.

> Nice, didnt know about TeaVMProblemRenderer. it gives more details!

Yes, there are plenty of things available in TeaVM that you didn't know of. That's why I often suggest my assistance with adopting standard Gradle plugin in gdx build pipeline instead of direct usage of `TeaVMTool`

**Reply by xpenatan** [reply]
Im investigating to add gradle teavm plugin support.  Using TeaVMTool manually I use TeaVMPlugin interface to do some internal logic. I think I can move some code from TeaBuilder to TeaPlugin class which implements TeaVMPlugim.  The only issue I will need to investigate is why using the gradle plugin is not calling TeaPlugin class.

**Reply by xpenatan** [reply]
Is there a way to get more information about this error?

![image](https://github.com/user-attachments/assets/c8edbaa8-eed1-43f1-8865-2a6b67ea49fc)


Edit. Toke me some time to get to the problem. I'm using JSByRef and that is why is failing.

**Reply by konsoletyper** [reply]
@xpenatan well done! Sorry for a bit weird error reporting, it should have be more user-friendly like "@JSByRef not supported by WebAssembly GC", but this is preview. As for investigation methods, you can use disassembler here: https://teavm.org/tools/wasm/disasm/. To make life easier, enable generation of `name` section in wasm file (it's the same option that controls minification of JS) and debug information (it should be embedded in wasm file, see `TeaVMTool.setWasmDebugInfoLocation`). In this case you'll get disassembly with all symbol names and comments referring to original source location (including inlining info).

**Reply by xpenatan** [reply]
@konsoletyper  I got the basic 3d gear demo to work.

I did something similar to what GWT backend do. I created a HasArrayBufferView interface for buffers so I can obtain the javascript object.

Test branch here https://github.com/xpenatan/teavm/commit/8d758d5290b2186837ca6710b086e5a5e33572ce (forgot to change dirty flag to false 😆)

For every buffer that contains java array I created a javascript typearray as a backup. If it the original array changes I mark a dirty flag so when getArrayBufferView is called It recreates the javascript array.   

I think if java array is replaced by javascript array a backup array wont be needed.  With HasArrayBufferView  interface I can obtain the javascript object and pass it to webgl.

![image](https://github.com/user-attachments/assets/470bbf2f-1df0-46a4-b355-bdd2d709c4c4)

**Reply by konsoletyper** [reply]
@xpenatan I don't think that writing data from Wasm to JS typed array directly is a good idea. In this case all calls pass through Wasm/JS barrier, which has some overhead. The proper solution is to write to Wasm heap and pass it to JS. Since JS sees Wasm heap as ArrayBuffer, this will work fast. Without any overhead. But I need first to support this mechanism.

**Reply by konsoletyper** [reply]
@xpenatan anyway, congrats! Hope you'll be able to make some performance and binary size comparison. Eventually, I'll implement remaining features, like reflection, and JS be will be completely replaced by Wasm in libGDX TeaVM BE. Also, small hint: you can try to disable strict mode, this should add some performance

**Reply by xpenatan** [reply]
> @xpenatan well done! Sorry for a bit weird error reporting, it should have be more user-friendly like "@JSByRef not supported by WebAssembly GC", but this is preview. As for investigation methods, you can use disassembler here: https://teavm.org/tools/wasm/disasm/. To make life easier, enable generation of `name` section in wasm file (it's the same option that controls minification of JS) and debug information (it should be embedded in wasm file, see `TeaVMTool.setWasmDebugInfoLocation`). In this case you'll get disassembly with all symbol names and comments referring to original source location (including inlining info).

Thanks.  The console error was not helping on which GL calls was the issue. The disarm helped 👍 
![image](https://github.com/user-attachments/assets/fcd5e4e0-846e-4033-ba7e-02b96768f6db)

![image](https://github.com/user-attachments/assets/e47459d8-e288-4b24-bc08-d43da1f10bac)

**Reply by konsoletyper** [reply]
@xpenatan see the address on first screenshot in the very end of error? It's address in decimal. Convert in to hex and search is disassembly. You'll get exact instruction

**Reply by xpenatan** [reply]
Pushing teavm to its limits with Metaprograming hacks and some others I got GLTF to work.

![image](https://github.com/user-attachments/assets/e1c0d043-1c75-498b-beab-cb8230c02ff9)

I will wait for proper reflection support and Wasm heap as ArrayBuffer mechanism.  Libgdx use Threads so I had to comment out some internal classes.

**Reply by konsoletyper** [reply]
@xpenatan great, here are my thoughts

1. LibGDX works with GWT, right? So there's at least some core subset which works without threads. I recommend to avoid using threads and `@Async` when possible, since it's performance killer. Moreover, it's *unpredictable* performance killer, which means that it does not affect performance of code that actually uses it, but of some code that you think is completely irrelevant.

2. As for ArrayBuffer, sure, I understand you. On the other hand, typical game should avoid uploading textures/meshes to GPU when possible and do this only when game/level loaded. Considering that Wasm could improve performance in other places (like game logic, physics, AI), somewhat slower loading could be reasonable trade off.

3. GWT lacks reflection, right? But it still works with libGDX. You can wait until I implement reflection (this would take several months or even a year), or do something with this right now. Why do you need reflection in the first place? IMO, reflection should be avoided when possible. In CoSpaces Edu we don't use reflection. To do JSON serializations and REST API we use Java annotation processors and Kapt/KSP respectively. Rationale: we compile both to Web, Android and iOS, and each platform has its own problems with reflection (on Android r8 needs special annotations, for iOS we use GraalVM, which also requires additional configuration). Also, reflection is a performance killer.

**Reply by xpenatan** [reply]
1:  Yes. It emulates some classes that GWT does not support. I'll take a look at it and do the same for teavm.

2: The whole copying buffer thing I got it wrong. I created a small profile to log all GL commands to see what is going on and buffers are only being used when loading models. The render loop is not passing any buffers to WebGL commands. 🙏 

3:  Yes, libgdx does by using GWT code generator.  I used metaprograming to get the last example working.  I'll start testing other demos to see how it goes.

**Comment by konsoletyper** [comment]
Hey, guys. I fixed several bugs according to feedback and published new preview build. I'm going for vacation, so I'll not be able to fix any further bugs for the next week.

**Comment by lax1dude** [comment]
Great job! I really appreciate you put in the time to get this all working. Please don’t kill off the legacy WASM backend yet though, my impression of WASM GC right now is that the authors if the WebAssembly spec rushed it, and it’s still less than even an MVP with their stupid arrays that can’t be copied efficiently. I don’t care that the legacy WASM backend makes a larger binary, I was gonna compress it with LZMA anyway which seems to reduce its size by like 75%. Wait until the WASM GC spec itself is actually complete before killing off the legacy WASM backend you spent nearly a decade working on.

**Comment by ScraM-Team** [comment]
Congratulations @konsoletyper!  I remember you were pushing the WASM folks to make WASM GC usable.  Thank you for your efforts to make TeaVM work with WASM (and WASM GC).

**Comment by StaticallyTypedAnxiety** [comment]
Congratulations @konsoletyper!

I am looking into actually updating wit-bindgen to work with 0.11.0 and I was wondering if there is a good place to look for memory operations? I am mostly looking for Malloc, Free and Put operations. I see that the WasiBuffer has a getBuffer operation, I was just wondering if there is a good place to start looking. 

Thank you!

**Comment by konsoletyper** [comment]
@StaticallyTypedAnxiety first of all, AFAIK, none of WASI runtimes support Wasm GC, so this thread is irrelevant for your question. Second, Java heap does not act in terms of "Malloc" and "Free". What's possible it to, for instance, allocate a byte array, but as soon as there won't be any reference to it from GC roots, it will be de-allocated on the next GC. Also, even if you manage to "pin" this array, it will be moved by next GC. Finally, there's no way to de-allocate array manually, it will be released by GC automatically. So there's no way to represent "Malloc" and "Free" operations.

**Comment by bertung** [comment]
İs there a more comprehensive example for wasm target showing more complex objects been exchanged between javascript and wasm code? such as strings or byte arrays.

**Reply by konsoletyper** [reply]
Isn't example that I linked in my post enough?

**Reply by bertung** [reply]
It does not deal with strings and byte arrays.

**Reply by konsoletyper** [reply]
[It does](https://github.com/konsoletyper/teavm/blob/7a2cf555e3d0eb05e25c7eb71eed21aa1db8176a/samples/benchmark/src/main/java/org/teavm/samples/benchmark/teavm/BenchmarkStarter.java#L60-L74) (at least with strings). With arrays - write your own example, it's all [documented](https://teavm.org/docs/runtime/jso.html)

**Reply by bertung** [reply]
Does that mean that what ever we do for javascript target of Teavm will also work with wasm target?

**Reply by konsoletyper** [reply]
@bertung it's exactly what this post states

**Comment by samyrjob** [comment]
We're in 2026 now and my favourite part of this post is you calling by-reference array passing impossible in theory before solving it in 0.12.0 anyway😄. Then Coroutines landed, the MVP backend got deleted, the compiler runs in the browser. Wild trajectory. Thanks for the huge job done, amazingly done , well done lad !

**Comment by konsoletyper** [comment]
> by-reference array passing impossible in theory before solving it in 0.12.0 anyway

By-reference array passing is not possible in Wasm GC, and it's still the issue. Also, it's not the issue of TeaVM implementation, but rather a limitation of Wasm GC spec. I believe, the only reason for this is "we don't have time for that". I also believe, the WebAssembly spec/support pace has significantly decreased last few years.

**Reply by samyrjob** [reply]
My bad, indeed after re-reading it,  Wasm GC arrays are still not viewable as typed arrays; TeaVM allocates direct NIO buffers in linear memory instead . Let's see whether the spec ever gets there.
